priv/posts/security/20241212210148_epmd-public-exposure.md

```diff
@@ -0,0 +1,114 @@
+- Many Erlang/Elixir build tools and release pipelines (like Rebar3, Mix, and Distillery releases) enable EPMD by default.
+
+## Why is Exposing EPMD a Bad Idea?
+
+### Risk of Unauthorized Cluster Access
+
+EPMD itself does not enforce strong authentication. Erlang clustering uses a “cookie” as a shared secret, but this cookie is not intended as a robust security mechanism—it’s more of a sanity check to prevent accidental cross-node connections. If an attacker can guess or brute-force the cookie, they could join your cluster.
+
+Once inside the cluster, the attacker could potentially run arbitrary Erlang Remote Procedure Calls (RPCs), giving them full control of the application and underlying system. While no widespread internet-based attacks on EPMD have been documented publicly, the theoretical risk is real. A known brute-force approach is described here:  
+[https://insinuator.net/2017/10/erlang-distribution-rce-and-a-cookie-bruteforcer/](https://insinuator.net/2017/10/erlang-distribution-rce-and-a-cookie-bruteforcer/)
+
+### Scope of the Problem
+
+Shodan (a search engine for internet-connected devices) reveals over 85,000 publicly accessible EPMD instances. Among them, around 40,000 are associated with RabbitMQ.  
+Top countries hosting exposed EPMD endpoints include China (18,444), the United States (15,771), and Germany (10,263). Major cloud hosting platforms (DigitalOcean, Hetzner, Aliyun, Tencent, OVH, Amazon, and others) also host significant numbers of exposed instances, typically due to default configurations and lack of firewall restrictions.
+
+[![World Map showing number of exposed EPMD instances](/images/posts/epmd-public-exposure/world-map.png) {: .responsive-image}](https://www.shodan.io/search/report?query=product%3A%22Erlang+Port+Mapper+Daemon%22)
+
+### Not Just Erlang Developers at Risk
+
+RabbitMQ, a popular message broker built on Erlang, runs EPMD under the hood. Many RabbitMQ users may not be aware that EPMD exists or that it’s exposed. This lack of awareness can lead to overlooked vulnerabilities, as users trust defaults without applying basic network hygiene.
+
+## How to Secure Your EPMD
+
+**1. Use a Firewall or Restricted Network Interfaces**
+
+If you don’t need external clustering, ensure EPMD isn’t exposed:
+
+- Bind EPMD to localhost or a private network interface. In `vm.args`, specify:
+  ```bash
+  -kernel inet_dist_use_interface '{127, 0, 0, 1}'
+  ```
+- Deploy firewalls, security groups, or network access control lists (ACLs) to ensure port 4369 is not reachable from untrusted networks.
+
+**2. Remove Clustering Options if Not Needed**
+
+If you’re not clustering multiple Erlang/RabbitMQ nodes, don’t start the node with `-name` or `-sname`. Without these options, your application won’t register with EPMD at all. If you only run a single RabbitMQ node and do not need distribution, consider disabling or limiting it according to the official networking and clustering documentation.
```

Comment:
```suggestion
If you’re not clustering multiple Erlang/RabbitMQ nodes, don’t start the node with `-name` or `-sname`. Without these options, your application won’t register with EPMD at all. If you only run a single RabbitMQ node and do not need distribution, consider disabling or limiting it according to the official networking and clustering documentation.
```

Many systems use `erl_call` or `remsh` for debugging tools, which requires distribution to be running. So even if you are not clustered to other machines, you may will need/want distrubution running for debugging.